docs: document struct coded and add examples (#406)
foundry-rs/book: src/misc/struct-encoding.md

> #### Struct Encoding

Reviewer: ```suggestion
## Struct Encoding
```

> Structs are custom defined types that can group several variables:

Reviewer: ```suggestion
Structs are user defined types that can group several variables:
```

```solidity
struct MyStruct {
    address addr;
    uint256 amount;
}
```

Only the new [ABI coder v2](https://docs.soliditylang.org/en/latest/layout-of-source-files.html#abi-coder-pragma) can encode and decode arbitrarily nested arrays and structs. Since Solidity 0.8.0 it is activated by default, prior to that it needs to be activated via `pragma experimental ABIEncoderV2`.

> Structs map to the ABI type "tuple", See also [Mapping Solidity to ABI types](https://docs.soliditylang.org/en/latest/abi-spec.html#mapping-solidity-to-abi-types)

Reviewer: ```suggestion
Solidity structs map to the ABI type "tuple". For more information on how Solidity types map to ABI types see [Mapping Solidity to ABI types](https://docs.soliditylang.org/en/latest/abi-spec.html#mapping-solidity-to-abi-types) in the Solidity documentation.
```

> Structs are therefor encoded and decodes as tuples. So `MyStruct` is a `(address,uint256)` tuple when it comes to the ABI spec.

Reviewer: ```suggestion
Structs are therefore encoded and decodes as tuples. So the struct we defined above, `MyStruct`, maps to the tuple `(address,uint256)` in terms of the ABI.
```

> For example the ABI of

Reviewer: ```suggestion
Let's see how this works in a contract:
```

```solidity
pragma solidity =0.8.15;

contract Test {
    struct MyStruct {
        address addr;
        uint256 amount;
    }
    function f(MyStruct memory t) public pure {}
}
```

> results in the following ABI json object:

Reviewer: ```suggestion
The ABI of the `f` function in this contract is:
```

```json
{
	"inputs": [
		{
			"components": [
				{
					"internalType": "address",
					"name": "addr",
					"type": "address"
				},
				{
					"internalType": "uint256",
					"name": "amount",
					"type": "uint256"
				}
			],
			"internalType": "struct Test.MyStruct",
			"name": "t",
			"type": "tuple"
		}
	],
	"name": "f",
	"outputs": [],
	"stateMutability": "pure",
	"type": "function"
}
```

> which reads:
> The function `f` takes 1 input of type `tuple` with two components of type `address` and `uint256`.

Reviewer: ```suggestion
which reads: The function `f` takes 1 input of type `tuple` with two components of type `address` and `uint256`.
```
